### PR TITLE
Trigger times added to Neuland

### PR DESCRIPTION
**Pull request by igasparic**
- in reader NeulandTrigMappedData + option fSkiptriggertime
- Mapped2CalPar
- Mapped2Cal - fTriggerTime in NeulandCalData - mapping_neuland_trig.hh hardcoded
- Cal2HitPar, Cal2Hit - trigger time subtracted

NeulandOnlineSpectra updated

GainMatching modified in preexp

Clang-format

### CODE REVIEW

File: neuland/calibration/R3BNeulandMapped2Cal.cxx [hunk 1]
```diff
@@ -24,14 +24,16 @@
 #include "TClonesArray.h"
 #include "TH2F.h"
 #include "TMath.h"
+#include "mapping_neuland_trig.hh"
```
Comment (jose-luis-rs) [inline neuland/calibration/R3BNeulandMapped2Cal.cxx:27]: I am wondering if this file depends on the experiment, if yes, it must be read from a parameter container. This kind of container was already created for other detectors like Fibers and TofD:
- https://github.com/R3BRootGroup/R3BRoot/blob/dev/tofd/pars/R3BTofDMappingPar.h
- https://github.com/R3BRootGroup/R3BRoot/blob/dev/fiber/pars/R3BFiberMappingPar.h

**Review by jose-luis-rs — CHANGES_REQUESTED** [review]

File: neuland/calibration/R3BNeulandMapped2Cal.cxx (hunk 1 above)
Comment (jose-luis-rs) [reply neuland/calibration/R3BNeulandMapped2Cal.cxx:27]: There are also some examples of macros to convert the ucesb mapping files into r3broot parameter files:
https://github.com/R3BRootGroup/R3BParams_S515/blob/dev/macros/calibration/fibers/make_fiber_trig_mapping.C
https://github.com/R3BRootGroup/R3BParams_S515/blob/dev/macros/calibration/tofd/make_tofd_trig_mapping.C

File: r3bdata/neulandData/R3BNeulandCalData.h [hunk 2]
```diff
@@ -21,11 +21,12 @@ class R3BNeulandCalData : public TObject
 {
   public:
     R3BNeulandCalData() = default;
-    R3BNeulandCalData(Int_t barId, Int_t side, Double_t time, Int_t qdc);
+    R3BNeulandCalData(Int_t barId, Int_t side, Double_t time, Double_t triggertime, Int_t qdc);
 
     Int_t GetBarId() const { return fBarId; }
     Int_t GetSide() const { return fSide; }
     Double_t GetTime() const { return fTime; }
```
Comment (YanzhaoW) [inline r3bdata/neulandData/R3BNeulandCalData.h:28]: Redefine GetTime method to get triggered time. Also add another getter for untriggered time:
```suggestion
    Double_t GetTime() const { return fTime - fTriggerTime; }
    Double_t GetTimeUntriggered() const { return fTime; }
```

File: r3bdata/neulandData/R3BNeulandCalData.cxx [hunk 3]
```diff
@@ -13,18 +13,20 @@
...
 
-R3BNeulandCalData::R3BNeulandCalData(Int_t barId, Int_t side, Double_t time, Int_t qdc)
+R3BNeulandCalData::R3BNeulandCalData(Int_t barId, Int_t side, Double_t time, Double_t triggertime, Int_t qdc)
     : fBarId(barId)
     , fSide(side)
     , fTime(time)
+    , fTriggerTime(triggertime)
     , fQdc(qdc)
 {
 }
```
Comment (YanzhaoW) [inline r3bdata/neulandData/R3BNeulandCalData.cxx:23]: Add a second constructor for the sake of backward compatibility.
```suggestion
R3BNeulandCalData::R3BNeulandCalData(Int_t barId, Int_t side, Double_t time, Double_t triggertime, Int_t qdc)
    : fBarId(barId)
    , fSide(side)
    , fTime(time)
    , fTriggerTime(triggertime)
    , fQdc(qdc)
{
}

R3BNeulandCalData::R3BNeulandCalData(Int_t barId, Int_t side, Double_t time, Int_t qdc)
    : R3BNeulandCalData(barId, side, time, 0.0, qdc)
{
}
```

File: r3bdata/neulandData/R3BNeulandCalData.h [hunk 4]
```diff
@@ -21,11 +21,12 @@ class R3BNeulandCalData : public TObject
 {
   public:
     R3BNeulandCalData() = default;
-    R3BNeulandCalData(Int_t barId, Int_t side, Double_t time, Int_t qdc);
+    R3BNeulandCalData(Int_t barId, Int_t side, Double_t time, Double_t triggertime, Int_t qdc);
```
Comment (YanzhaoW) [inline r3bdata/neulandData/R3BNeulandCalData.h:24]:
```suggestion
    R3BNeulandCalData(Int_t barId, Int_t side, Double_t time, Double_t triggertime, Int_t qdc);
    R3BNeulandCalData(Int_t barId, Int_t side, Double_t time, Int_t qdc);
```

File: neuland/calibration/R3BNeulandCal2Hit.cxx [hunk 5]
```diff
@@ -171,8 +171,17 @@ void R3BNeulandCal2Hit::Exec(Option_t*)
...
+
+        if (std::isnan(cal[0]->GetTriggerTime()) || std::isnan(cal[0]->GetTriggerTime()))
+        {
+            tdc = { cal[0]->GetTime() + parameter.GetTimeOffset(1), cal[1]->GetTime() + parameter.GetTimeOffset(2) };
+        }
+        else
+        {
+            tdc = { cal[0]->GetTime() - cal[0]->GetTriggerTime() + parameter.GetTimeOffset(1),
+                    cal[1]->GetTime() - cal[1]->GetTriggerTime() + parameter.GetTimeOffset(2) };
+        }
```
Comment (YanzhaoW) [inline neuland/calibration/R3BNeulandCal2Hit.cxx:184]: Such a change in unnecessary if we just set the default value of triggerTime to 0.

File: neuland/calibration/R3BNeulandCal2HitPar.cxx [hunk 6]
```diff
@@ -128,7 +128,14 @@ void R3BNeulandCal2HitPar::Exec(Option_t* option)
...
 
-        fHitCalEngine->Set(id, side, pmt->GetTime(), pmt->GetQdc());
+        if (std::isnan(pmt->GetTriggerTime()))
+        {
+            fHitCalEngine->Set(id, side, pmt->GetTime(), pmt->GetQdc());
+        }
+        else
+        {
+            fHitCalEngine->Set(id, side, pmt->GetTime() - pmt->GetTriggerTime(), pmt->GetQdc());
+        }
```
Comment (YanzhaoW) [inline neuland/calibration/R3BNeulandCal2HitPar.cxx:138]: Also not necessary if the default value of triggerTime is 0.

**Review by YanzhaoW — COMMENTED** [review]
I think there is no necessity to change anything in Cal2Hit file if we redefine the `GetTime()` to the triggered time value.

**Review by YanzhaoW — COMMENTED** [review]
Some suggestions in MappedToCal file.

**Comment by igasparic** [comment]
Thanks for the suggestions. A new pull request is created but only with fixes related to Jose Luis' comments. I will take care of Yanzhao's comments in future.

**Comment by jose-luis-rs** [comment]
Hi
This PR and #664 are the same, right?
Can I close this one?
Cheers

**Comment by igasparic** [comment]
Yes, close this one. #664 contains everything from here and R3BNeulandMappingPar stuff in addition.

**Pull request closed without merging**